File: leetcode.com/0515_find_largest_value_in_each_tree_row/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    X = -2**32

    def trace(self, node, idx):
        if not node:
            return

        self.rows[idx] = max(self.rows[idx], node.val)
        self.trace(node.left, idx + 1)
        self.trace(node.right, idx + 1)

    def largestValues(self, root: TreeNode) -> List[int]:
        self.rows = [self.X for _ in range(10**4)]
        self.trace(root, 0)
        return filter(lambda v: v is not self.X, self.rows)

File: leetcode.com/0515_find_largest_value_in_each_tree_row/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 func max(a, b int) int {
    if (a > b) {
        return a
    }

    return b
}

func largestValues(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }

    values := []int{}
    queue := []*TreeNode{root}

    for len(queue) > 0 {
        val := queue[0].Val
        next_queue := []*TreeNode{}

        for _, item := range queue {
            val = max(val, item.Val)

            if item.Left != nil {
                next_queue = append(next_queue, item.Left)
            }

            if item.Right != nil {
                next_queue = append(next_queue, item.Right)
            }
        }

        values = append(values, val)
        queue = next_queue
    }

    return values
}